src/cart/crt_swim.c

```diff
@@ -742,29 +762,60 @@ static int64_t crt_swim_progress_cb(crt_context_t crt_ctx, int64_t timeout, void
 	int			 rc;
 
 	if (self_id == SWIM_ID_INVALID)
-		return timeout;
+		return timeout_us;
 
 	if (crt_swim_fail_hlc && crt_hlc_get() >= crt_swim_fail_hlc) {
 		crt_swim_should_fail = true;
 		crt_swim_fail_hlc = 0;
 		D_EMIT("SWIM id=%lu should fail\n", crt_swim_fail_id);
 	}
 
-	rc = swim_progress(ctx, timeout);
+	rc = swim_progress(ctx, timeout_us);
 	if (rc == -DER_SHUTDOWN) {
 		if (grp_priv->gp_size > 1)
 			D_ERROR("SWIM shutdown\n");
 		swim_self_set(ctx, SWIM_ID_INVALID);
 	} else if (rc == -DER_TIMEDOUT || rc == -DER_CANCELED) {
-		uint64_t now = swim_now_ms();
+		crt_swim_update_last_unpack_hlc(csm);
 
-		if (now < ctx->sc_next_event)
-			timeout = ctx->sc_next_event - now;
+		/*
+		 * Check for network idle in all contexts.
+		 * If the time passed from last received RPC till now is more
+		 * than 2/3 of suspicion timeout suspends eviction.
+		 * The max_delay should be less suspicion timeout to guarantee
+		 * the already suspected members will not be expired.
+		 */
+		if (csm->csm_alive_count > 2) {
+			uint64_t hlc1 = csm->csm_last_unpack_hlc;
+			uint64_t hlc2 = crt_hlc_get();
+			uint64_t delay = crt_hlc2msec(hlc2 - hlc1);
+			uint64_t max_delay = swim_suspect_timeout_get() * 2 / 3;
+
+			if (delay > max_delay) {
+				D_ERROR("Network outage detected (idle during "
+					"%lu.%lu sec > expected %lu.%lu sec).\n",
+					delay / 1000, delay % 1000,
+					max_delay / 1000, max_delay % 1000);
+				swim_net_glitch_update(csm->csm_ctx, self_id, delay);
+				csm->csm_last_unpack_hlc = hlc2;
+			}
+		}
+
+		/*
+		 * Change only for very long timeout to avoid confusing of DAOS scheduler.
+		 * NB: !!! Mercury only supports milli-second timeout !!!
+		 */
+		if (timeout_us > 1000) {
+			uint64_t now = swim_now_ms();
+
+			if (now < ctx->sc_next_event)
+				timeout_us = (ctx->sc_next_event - now) * 1000; /* timeout in us */
```

Comment:
If crt_progress is called with timeout being 2000 us, we don't want crt_swim_progress_cb to return a timeout > 2000 us, right? Can the new timeout_us value > the original timeout_us value?

Reply:
The idea of this timeout change was accommodate to real events progressing. If nothing to progress we should increase timeout and spend all time inside network layer. If we expect near event we should ask network layer to return sooner. So, here I estimate the time to next event and pass corresponding value to network layer. And yes, new timeout can be bigger than original.

Reply:
I think crt_swim_progress_cb need to make sure the returned timeout_us is <= the original timeout_us.

@frostedcmos, @NiuYawei, any comments?

Reply:
@dmiter what would be a valid case when a modified timeout exceeds the original timeout? 
it seems to me that the original timeout should be the upper limit value, and then we can adjust up/down based on network demand, but the original value should be an upper bound

Reply:
TBH, I don't quite follow this change. Why it would confuse DAOS scheduler (if we don't enlarge timeout here)? And why it's related to "Mercury only supports milli-second timeout"?

Anyway, for SWIM xstream, scheduler should think it's always 'idle', so scheduler will always pass 1ms timeout (default value) to crt_progress().